utils/s3/client.cc

```diff
@@ -98,6 +98,191 @@ future<> client::delete_object(sstring object_name) {
+    _part_etags.emplace_back();
+    co_await upload_part(pn, std::move(_bufs));
+}
+
+sstring parse_multipart_upload_id(sstring& body) {
+    rapidxml::xml_document<> doc;
+    std::exception_ptr ex;
+    try {
+        doc.parse<0>(body.data());
+    } catch (const rapidxml::parse_error& e) {
+        s3l.warn("cannot parse initiate multipart upload response: {}", e.what());
+        return "";
+    }
+    auto root_node = doc.first_node("InitiateMultipartUploadResult");
+    auto uploadid_node = root_node->first_node("UploadId");
+    return uploadid_node->value();
+}
+
+sstring dump_multipart_upload_parts(const utils::chunked_vector<sstring>& etags) {
+    sstring body("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\r\n");
+    body += sstring("<CompleteMultipartUpload xmlns=\"http://s3.amazonaws.com/doc/2006-03-01/\">");
+    unsigned nr = 0;
+    for (auto& etag : etags) {
+        if (etag.empty()) {
+            return "";
+        }
+
+        body += format("<Part><ETag>{}</ETag><PartNumber>{}</PartNumber></Part>", etag, nr++);
+    }
+    return body + sstring("</CompleteMultipartUpload>");
+}
+
+future<> client::upload_sink::start_upload() {
+    s3l.trace("POST uploads {}", _object_name);
+    auto rep = http::request::make("POST", _client->_host, _object_name);
+    rep.query_parameters["uploads"] = "";
+    co_await _http.make_request(std::move(rep), [this] (const http::reply& rep, input_stream<char>&& in_) -> future<> {
+        auto in = std::move(in_);
+        auto body = co_await util::read_entire_stream_contiguous(in);
+        _upload_id = parse_multipart_upload_id(body);
+        if (_upload_id.empty()) {
+            co_await coroutine::return_exception(std::runtime_error("cannot initiate upload"));
+        }
+        s3l.trace("created uploads for {} -> id = {}", _object_name, _upload_id);
+    });
+}
+
+future<> client::upload_sink::upload_part(unsigned part_number, memory_data_sink_buffers bufs) {
+    s3l.trace("PUT part {} {} bytes in {} buffers (upload id {})", part_number, bufs.size(), bufs.buffers().size(), _upload_id);
+    auto req = http::request::make("PUT", _client->_host, _object_name);
+    req._headers["Content-Length"] = format("{}", bufs.size());
+    req.query_parameters["partNumber"] = format("{}", part_number);
+    req.query_parameters["uploadId"] = _upload_id;
+    req.write_body("bin", bufs.size(), [this, part_number, bufs = std::move(bufs)] (output_stream<char>&& out_) mutable -> future<> {
+        auto out = std::move(out_);
+        s3l.trace("upload {} part data (upload id {})", part_number, _upload_id);
+        for (auto&& buf : bufs.buffers()) {
+            co_await out.write(buf.get(), buf.size());
+        }
+        co_await out.flush();
+        co_await out.close();
+    });
+
+    auto units = co_await get_units(_flush_sem, 1);
+    (void)_http.make_request(std::move(req), [this, part_number] (http::reply rep, input_stream<char>&& in_) mutable -> future<> {
```

Comment:
Need comment justifying ignored future and explaining error handling.

Reply:
I am wondering why we do not wait for make_request in client::upload_sink::upload_part in the first place?

Reply:
To issue upload parts in parallel. It will wait in `get_units()` if runs out of parallelizm